### PR TITLE
feat(store): add atomic find and replace updates

### LINKED ISSUE #602: feat(store): add find+replace params to mem_update for token-efficient partial content edits
### 📋 Pre-flight Checks
- [x] I have searched existing issues and this is not a duplicate
- [x] I understand this issue needs status:approved before a PR can be opened

### 🔍 Problem Description

Editing a memory to fix a typo, update a file path, or rename a concept currently requires sending the entire content block. This wastes tokens and adds round-trips — especially painful for AI agents.

To fix a single word, the agent must either:
- (a) remember the full content and re-send it, or
- (b) call mem_get_observation first, then re-send the full body with the correction

Neither is token-efficient.

### 💡 Proposed Solution

Add two optional, mutually-dependent string parameters to `mem_update`: `find` and `replace`.

When both are provided (neither alone is valid), the content field is modified in-place by replacing ALL occurrences of `find` with `replace`. This happens server-side without sending the full content.

Implementation: in `internal/store/store.go:UpdateObservation`, after the existing observation is read (which already happens in the transaction), apply `strings.ReplaceAll` to the content before the UPDATE query. The existing truncation, hash, revision_count, and sync logic apply automatically.

**Example:**
```
mem_update(
  id: 42,
  find: "mnemonic",
  replace: "memonic"
)
```

**Validation rules:**
- `find` without `replace` → error: "find and replace must be used together"
- `replace` without `find` → error: "find and replace must be used together"
- `find` + `replace` with `content` → error: "find/replace is mutually exclusive with content"
- `find` is empty string → no-op
- `find` not found → no-op, no error

### 📦 Affected Area

- Store (database, queries)
- MCP Server (tools, transport)

### 🔄 Alternatives Considered

**Client-side approach:** Read via mem_get_observation, modify locally, re-send via mem_update. Adds a round-trip and still sends the full content.

**SQL-level REPLACE:** Using SQLite REPLACE(content, ?, ?) directly. Viable but complicates content truncation logic currently handled in Go.

**JSON string parameter:** Single `find_replace` field as JSON. Rejected — cannot be validated by MCP tool schema, breaks flat-parameter pattern used by all other mem_update fields.

### 📎 Additional Context

**Token savings (1 token / 4 chars, conservative):**

| Scenario | Current | With find/replace | Savings |
|---|---|---|---|
| Typo fix in ~300 char memory | ~95 tok | ~16 tok | ~79 (83%) |
| Path update in ~800 char memory | ~220 tok | ~16 tok | ~204 (93%) |
| Concept rename × N memories | ~150tok × N | ~20tok × N | ~130-200tok per edit |

**Implementation reference files:**
- Handler: `internal/mcp/mcp.go:1055-1110`
- Store: `internal/store/store.go:2349-2420`
- Params struct: `internal/store/store.go:149-156`

### PR DESCRIPTION
**Pull request by blak0p**
## 🔗 Linked Issue

Closes #602

---

## 🏷️ PR Type

- [x] `type:feature` — New feature
- [ ] `type:bug` — Bug fix
- [ ] `type:docs` — Documentation only
- [ ] `type:refactor` — Code refactoring
- [ ] `type:chore` — Maintenance, dependencies, tooling
- [ ] `type:breaking-change` — Breaking change

---

## 📝 Summary

- Add atomic literal find/replace updates through MCP and HTTP PATCH.
- Preserve validation, normalization, revision, and sync semantics.
- Document the feature and add store, transport, and binary-backed coverage.

## 📂 Changes

| File | Change |
|------|--------|
| `internal/store/*` | Atomic replacement, validation, and persistence tests |
| `internal/mcp/*` | MCP schema and handler support |
| `internal/server/*` | HTTP PATCH behavior and contract tests |
| `DOCS.md` | User-facing API documentation |

## 🧪 Test Plan

- [x] Unit tests: `go test ./...`
- [x] E2E tests: `go test -tags e2e ./internal/server/...`
- [x] `go vet ./...` and formatting checks
- [x] Real binary against isolated SQLite: valid replacement, empty-find no-op, and invalid requests

---

## ✅ Contributor Checklist

- [x] Linked approved issue #602
- [x] Added exactly one `type:*` label
- [x] Ran unit and E2E tests
- [x] Updated documentation
- [x] Used a conventional commit with no `Co-Authored-By` trailer

## 💬 Notes for Reviewers

The change is intentionally one cohesive PR under an approved size exception because the store contract, both transports, and persistence proof share a single behavior boundary.

## Summary by CodeRabbit

* **New Features**
  * Added find-and-replace updates for observation content.
  * Supports replacing all literal matches while preserving content limits and privacy redaction.
  * Updates remain synchronized across cloud imports and local revisions.

* **Bug Fixes**
  * Added validation for incomplete, oversized, conflicting, or empty find-and-replace inputs.
  * Ensured replacements are safely truncated without breaking UTF-8 text.

### CODE REVIEW

File: internal/store/store.go [hunk 1]
```diff
@@ -2368,7 +2379,15 @@ func (s *Store) UpdateObservation(id int64, p UpdateObservationParams) (*Observa
...
+		if p.Find != nil {
+			if *p.Find != "" {
+				content = strings.ReplaceAll(content, *p.Find, *p.Replace)
+			}
+		}
+		if p.Content != nil || p.Find != nil {
+			content = stripPrivateTags(content)
 			if len(content) > s.cfg.MaxObservationLength {
 				content = content[:s.cfg.MaxObservationLength] + "... [truncated]"
 			}
```
[the remark this answers is not in the document]
Comment (blak0p) [reply internal/store/store.go]: Fixed in 877ce78. The replacement path now bounds retained output without materializing the expanded string, while preserving the existing sanitization and truncation behavior. Added regression coverage for oversized repeated replacements and sanitization-equivalence edge cases. Full tests, vet, and E2E pass.

**Review by dnlrsls — CHANGES_REQUESTED** [review]
Thanks for implementing the full store/MCP/HTTP path and for addressing the original expanded-allocation finding. I reviewed exact head `877ce784cab440521ec26181bc82b9e7651c6728` against current `main@70870987f56ae662acf3cfa36fa7e9a99efa9fb3`. The focused store, MCP, and server tests pass on the candidate head, and `git diff --check` is clean. This still needs changes before approval.

## 1. Rebase semantically onto current main

The synthetic merge conflicts in:

- `DOCS.md`
- `internal/mcp/mcp_test.go`
- `internal/server/server.go`
- `internal/server/server_test.go`
- `internal/store/store.go`

This cannot be resolved by mechanically taking either side. The implementation predates current title/content admission, observation-project immutability, UTF-8-safe truncation, truncation metadata, sync behavior, MCP project-ownership checks, and response-envelope handling. Please reimplement the feature on current main while preserving all of those contracts, especially the ownership checks in the current `handleUpdate` path.

## 2. Preserve observation validity and exact no-op semantics

The current transform can replace the complete content with `""` or whitespace/private-only output and then persist an invalid observation. Validate the post-replacement, post-redaction result and roll the transaction back with the current empty-content error.

An empty or non-matching `find` is documented as a content no-op, but the candidate still sanitizes/truncates the existing content. Preserve the stored content byte-for-byte in those cases while retaining the explicitly designed revision and sync side effects.

The bounded writer also clips raw bytes and can split a multi-byte rune. Reuse current main's UTF-8-safe truncation semantics and add boundary coverage.

## 3. Bound work, not only retained output

`replaceAndNormalizeObservationContent` bounds the bytes retained in memory, but it continues walking every byte of every replacement for every match after the output is already truncated. Because `replace` is externally supplied and PATCH decoding is not bounded here, a modest request can amplify into very large CPU work.

Please add an early-stop/bounded-work design that remains correct around private-tag state, apply an appropriate request/input bound, and cover the amplification case. The existing 32-match × 256 KiB test proves the final output, not that work stops at the storage boundary.

## 4. Keep MCP and documentation contracts aligned

- Report replacement-driven truncation through the current MCP truncation metadata/warning path; `contentLen` currently covers direct content only.
- Keep current project-resolution/ownership checks and response envelope behavior.
- Do not document `project` as a public `mem_update` field: the public schema does not expose it, and observation project identity is immutable on current main.
- Preserve the current tool counts and other documentation changes added since this branch diverged.

Required regression coverage should include replacement-to-empty rejection, exact empty/non-match content no-op, UTF-8 truncation boundaries, enrolled and unenrolled sync behavior, MCP ownership/envelope/truncation metadata, HTTP/MCP omitted/empty/null/malformed combinations, literal regex characters, overlapping matches, and bounded-work stress.

## Review size

The current diff is 769 additions + 13 deletions across seven files, above the 400-line review budget, and I could not find an approved `size:exception`. Please make one honest split: first the current-main store contract with its behavior tests, then MCP/HTTP/docs with their transport tests. If the corrected core cannot be split safely after that pass, request an explicit maintainer size exception with the final line count rather than separating tests from the behavior they protect.

**Comment by blak0p** [comment]
The core recovery is now in this PR at `e2afcb8`, and the independently verified MCP/HTTP/docs slice is ready at `fd16459` on my fork.

To keep the review split, could a maintainer create `feat/mem-update-find-replace` in `Gentleman-Programming/engram` at `e2afcb8`? GitHub requires the base of a stacked upstream PR to be an upstream branch, while #774's head currently lives in my fork. Once that base exists, I can open the transport PR against it and retarget it to `main` after #774 merges.

**Comment by dnlrsls** [comment]
Thanks for taking this on and for working through the store, MCP, HTTP, truncation, and sync edge cases. We refined the contract on #602 and opened #1342 as a fresh current-main implementation so the shared behavior can be reviewed as one coherent slice.\n\nTo avoid maintaining two competing implementations for the same approved issue, I’m closing this PR as superseded by #1342. Your work helped surface several of the safety cases captured in the final design, especially bounded growth and truncation behavior.
